src/fiber/channel.rs

```diff
@@ -1006,33 +1007,65 @@ where
                 }
             }
 
-            // if this is the last hop, store the preimage.
-            // though we will RemoveTlcFulfill the TLC in try_to_settle_down_tlc function,
-            // here we can do error check early here for better error handling.
-            let preimage = peeled_onion_packet
-                .current
-                .payment_preimage
-                .or_else(|| self.store.get_invoice_preimage(&add_tlc.payment_hash));
-
-            if let Some(preimage) = preimage {
-                let filled_payment_hash: Hash256 = add_tlc.hash_algorithm.hash(preimage).into();
-                if add_tlc.payment_hash != filled_payment_hash {
-                    return Err(ProcessingChannelError::FinalIncorrectPreimage);
-                }
-                // update invoice status to received only all the error checking passed
-                if let Some(_invoice) = self.store.get_invoice(&payment_hash) {
-                    self.store
-                        .update_invoice_status(&payment_hash, CkbInvoiceStatus::Received)
-                        .expect("update invoice status failed");
+            let preimage = match peeled_onion_packet.current.payment_preimage {
+                None => {
+                    match self.store.get_invoice_preimage(&payment_hash) {
+                        Some(preimage) => preimage,
+                        None => {
+                            let status = self
+                                .store
+                                .get_invoice_status(&payment_hash)
+                                // The sender sent a TLC with no invoice associated with it.
+                                .ok_or(ProcessingChannelError::FinalIncorrectPaymentHash)?;
+                            let is_active = status == CkbInvoiceStatus::Open
+                                || status == CkbInvoiceStatus::Received;
+                            let is_settled =
+                                self.store.get_invoice_preimage(&payment_hash).is_some();
+                            if is_active && !is_settled {
+                                // This TLC is added to applied_add_tlcs in above, but
+                                // TLCs in the list applied_add_tlcs wouldn't be processed again.
+                                // For the unsettled active hold invoice TLCs, we should process them indefinitely
+                                // until they expire or are settled.
+                                state.tlc_state.applied_add_tlcs.remove(&add_tlc.tlc_id);
```

Comment:
`applied_add_tlcs` is introduced to resolve some concurrency issue, I'm not sure is it ok to remove it from here.

Reply:
I think so. Because we are actually adding and removing the tlc from exactly the same function `apply_add_tlc_operation_with_peeled_onion_packet`.